README.md
# Flowcase

Copyright 2012, 2013 Colin Wetherbee <cww@cpan.org>

Flowcase is a dynamic web page slideshow.

Using Perl and Template::Toolkit, you can create any dynamic content you want
and have it automatically refresh in a browser.  This is useful for hall
displays, conference kiosk boards, and any other context in which you might
have many different pieces of information to display, over time, to an
audience.

Flowcase was originally developed in order to display the current weather and
a stock market snapshot on a monitor in the author's home office, but it was
also designed to be scalable to serve any number of pages the user desires.

## Usage

For command-line options:

    Flowcase$ perl -Ilib bin/flowcase --help

Flowcase modules can be configured with `-c` options from the command line.
For example, in order to use the weather module, one must sign up for an API
key at wunderground.com and specify it, along with a latitude and longitude
(or another Weather Underground location identifier) on the command line:

    Flowcase$ perl -Ilib bin/flowcase -t ../pages \
        -c Weather.place=40.446808,-79.940140 -c Weather.apikey=abcdef01234


Flowcase/Changes
Revision history for Flowcase

0.01    Date/time
        First version, released on an unsuspecting world.



Flowcase/Makefile.PL
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Flowcase',
    AUTHOR              => q{Colin Wetherbee <cww@cpan.org>},
    VERSION_FROM        => 'lib/Flowcase.pm',
    ABSTRACT_FROM       => 'lib/Flowcase.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'gpl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Flowcase-*' },
);


Flowcase/lib/Flowcase/PageFile.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase::PageFile;

use common::sense;

use Moose;

use overload q{""} => sub
{
    my ($self) = @_;
    return $self->classname();
};

has 'filename' => (is => 'ro', isa => 'Str');
has 'pagename' => (is => 'ro', isa => 'Str');
has 'classname' => (is => 'ro', isa => 'Str');
has 'module' => (is => 'ro', isa => 'Str');

sub BUILD
{
    my ($self) = @_;

    my $page = $self->pagename();
    $page =~ s/\..*$//;
    $page =~ s/^(.)/uc($1)/e;
    $self->{classname} = $page;
    $self->{module} = Flowcase::get_instance()->module_path() . '::' . $page;
}

1;


Flowcase/lib/Flowcase/PageCache.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase::PageCache;

use common::sense;

use Moose;
use Template;
use Template::Constants qw(:debug);

use Flowcase;

sub BUILD
{
    my ($self) = @_;

    $self->{template_processor} = Template->new
    (
        { INCLUDE_PATH => Flowcase::get_instance()->tmpl_dir() },
    );
    $self->{page_cache} = { };
}

sub get_page
{
    my ($self, $page_file) = @_;

    say "Handling page [$page_file]";

    my $filename = $page_file->filename();

    # Serve non-template files directly.
    if ($filename !~ /\.tmpl$/)
    {
        my $fh;
        if (!open($fh, '<', $filename))
        {
            say "Unable to open file [$filename]: $!";
            return undef;
        }
        my $content = q{};
        $content .= $_ while ($_ = <$fh>);
        close($fh);
        return $content;
    }

    my $page = $page_file->pagename();

    my $cached_text = $self->_get_cached_page($page);
    if (defined($cached_text))
    {
        say 'Using cached content';
        return $cached_text;
    }

    my $bare_module = $page_file->classname();
    my $module = $page_file->module();

    say "Loading module [$module]";
    eval "require $module";
    if ($@)
    {
        say "Unable to load module [$module]: $@";
        return undef;
    }

    my $config_ref = Flowcase::get_instance()->config();
    my %module_opts;
    for my $key (%$config_ref)
    {
        if ($key =~ /^$bare_module/)
        {
            my $value = $config_ref->{$key};
            $key =~ s/^.*?\.//;
            $module_opts{$key} = $value;
        }
    }
    my $obj;
    my $vars_ref;
    eval
    {
        $obj = "$module"->new(\%module_opts);
        $vars_ref = $obj->get_vars();
    };
    if ($@)
    {
        say "Unable to retrieve variables from module [$module]: $@";
        return undef;
    }

    $vars_ref->{flowcase} =
    {
        page_delay => Flowcase::get_instance()->page_delay(),
    };

    my $text = q{};
    my $template = $self->{template_processor};
    $template->process("$page.tmpl", $vars_ref, \$text);

    my $error = $template->error();
    if ($error)
    {
        say "TT ERROR: " . $error;
    }

    my $cache_ttl;
    eval
    {
        $cache_ttl = $obj->cache_ttl();
    };
    if ($@)
    {
        $cache_ttl = 0;
    }

    if ($cache_ttl > 0)
    {
        say "Caching page for [$cache_ttl] seconds";
        $self->_set_cached_page($page, $text, $cache_ttl);
    }

    return $text;
}

sub _get_cached_page
{
    my ($self, $page) = @_;

    my $cache = $self->{page_cache}->{$page};
    if ($cache)
    {
        if (time() > $cache->{cache_time} + $cache->{ttl})
        {
            # Value is expired; delete it.
            say "Deleting expired cache content for page [$page]";
            delete $self->{page_cache}->{$page};
            return undef;
        }

        return $cache->{content};
    }

    return undef;
}

sub _set_cached_page
{
    my ($self, $page, $content, $ttl) = @_;

    $self->{page_cache}->{$page} =
    {
        cache_time => time(),
        content => $content,
        ttl => $ttl,
    };
}

sub _is_valid_page
{
    my ($self, $page) = @_;

    my $flowcase = Flowcase::get_instance();
    my $tmpl_dir = $flowcase->tmpl_dir();
    my $tmpl_file = "$tmpl_dir/$page.tmpl";
    say "Checking for existence of file [$tmpl_file]";
    return -e $tmpl_file ? $tmpl_file : undef;
}

1;


Flowcase/lib/Flowcase/WebHandler.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase::WebHandler;

use common::sense;

use File::Glob;
use HTTP::Daemon;
use HTTP::Status qw();
use List::Util qw(shuffle);
use Moose;
use Time::HiRes;

use Flowcase;
use Flowcase::PageCache;
use Flowcase::PageFile;

use constant ACCEPT_TIMEOUT => 2;

sub start
{
    my ($self) = @_;

    local $| = 1;
    
    my $server = $self->_start_server();

    my $cache = Flowcase::PageCache->new();
    my $finished = 0;

    my $flowcase = Flowcase::get_instance();

    my $bad_files_ref = $self->_load_file_lists($flowcase->exclude());
    if (scalar(@$bad_files_ref) > 0)
    {
        die 'Bad files in template directory: ',
            join(q{, }, @$bad_files_ref);
    }
    elsif (scalar(@{$self->{page_files}}) == 0)
    {
        die 'No files to serve in template directory: ',
            $flowcase->tmpl_dir();
    }

    my $cycle_page_files_ref = [ shuffle(@{$self->{page_files}}) ];
    my $cur_page = $cycle_page_files_ref->[0];
    my $cycle_timer = time();
    say "Active page is now [$cur_page]";

    local $SIG{INT} = sub
    {
        say 'Interrupt!';
        $finished = 1;
    };

    say 'Listening for new connections';

    while (!$finished)
    {
        if (time() > $cycle_timer + Flowcase::get_instance()->page_delay())
        {
            $cur_page = pop(@$cycle_page_files_ref);
            unshift(@$cycle_page_files_ref, $cur_page);
            $cycle_timer = time();
            say "Active page is now [$cur_page]";
        }

        my $c = $server->accept();
        next unless $c;

        while (my $r = $c->get_request())
        {
            if ($r->method() eq 'GET')
            {
                my $url = $r->uri()->path();
                say "Request for URL [$url]";
                $url =~ s|^/||;
                my $page_to_load;

                if ($url eq q{})
                {
                    # Send the current page in the slideshow.
                    $page_to_load = $cur_page;
                }
                else
                {
                    # Check whether a static file by the requested name
                    # exists.  If so, send it.
                    for my $file (@{$self->{static_files}})
                    {
                        if ($file->pagename() eq $url)
                        {
                            $page_to_load = $file;
                            last;
                        }
                    }
                }

                if (defined($page_to_load))
                {
                    my $resp_page = $cache->get_page($page_to_load);
                    if (defined($resp_page))
                    {
                        my $len = length($resp_page);
                        say "Response: full page ($len characters)";
                        my $response = HTTP::Response->new();
                        $response->code(HTTP::Status::HTTP_OK);
                        $response->content($resp_page);
                        $response->header('Connection' => 'close');
                        $c->send_response($response);
                    }
                    else
                    {
                        say 'Response: 404 (non-existent page)';
                        $c->send_error(HTTP::Status::HTTP_NOT_FOUND);
                    }
                }
                else
                {
                    say 'Response: 404 (bad path)';
                    $c->send_error(HTTP::Status::HTTP_NOT_FOUND);
                }
            }
            else
            {
                say 'Response: 405 (not a GET request)';
                $c->send_error(HTTP::Status::HTTP_METHOD_NOT_ALLOWED);
            }
        }

        say 'Closing connection';
        $c->close();
        undef $c;
    }

    say 'Shutting down';
}

sub _load_file_lists
{
    my ($self, $exclude_ref) = @_;

    my $tmpl_dir = Flowcase::get_instance()->tmpl_dir();
    my $file_spec = "$tmpl_dir/*";
    my @files = File::Glob::glob($file_spec);
    my @errors;
    $self->{page_files} = [];
    $self->{static_files} = [];
    for my $file (@files)
    {
        given($file)
        {
            when (m|/([^/_]*)\.tmpl|)
            {
                my $pagename = $1;

                # Work-around for Perl bug #94682, which breaks the ability to
                # use $_ inside a given/when clause.  I decided to rewrite
                # List::MoreUtils::any() here so that I didn't have to require
                # users to use Perl 5.15.3, in which that bug is fixed.
                my $excluded = 0;
                for my $exclude_file (@$exclude_ref)
                {
                    if ($pagename eq $exclude_file)
                    {
                        $excluded = 1;
                        last;
                    }
                }

                if (!$excluded)
                {
                    my $o = Flowcase::PageFile->new
                    (
                        filename => $file,
                        pagename => $pagename,
                    );
                    push(@{$self->{page_files}}, $o);
                }
            }
            when (m|/(_[^/]*)$| && !m|\.tmpl$|)
            {
                my $o = Flowcase::PageFile->new
                (
                    filename => $file,
                    pagename => $1,
                );
                push(@{$self->{static_files}}, $o);
            }
            when (m|_[^/]*\.tmpl$|)
            {
                # No op.  These files are used internally as TT include
                # files.
            }
            default
            {
                push(@errors, $file);
            }
        }
    }

    say 'Page files: ', join(q{, }, @{$self->{page_files}});
    say 'Static files: ', join(q{, }, @{$self->{static_files}});

    return \@errors;
}

sub _start_server
{
    my $flowcase = Flowcase::get_instance();

    my $server = HTTP::Daemon->new
    (
        LocalAddr => $flowcase->listen_address(),
        LocalPort => $flowcase->listen_port(),
        Timeout   => ACCEPT_TIMEOUT,
    ) || die 'Unable to open listen port';

    return $server;
}

1;



Flowcase/lib/Flowcase/Pages/Hello.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase::Pages::Hello;

use common::sense;

use Moose;

sub get_vars
{
    return { place => 'world' };
};

1;


Flowcase/lib/Flowcase/Pages/Market.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase::Pages::Market;

use common::sense;

use Finance::Quote;
use Moose;

use constant DEFAULT_HEADLINE_SYMBOL => '^GSPC';
use constant DEFAULT_SYMBOL_LIST => 'CNX,IVR,AGNC,BRKB,GLD';
use constant QUOTE_LOCATION => 'usa';

has 'headline_symbol' =>
    (is => 'ro', isa => 'Str', default => DEFAULT_HEADLINE_SYMBOL);
has 'symbol_list' =>
    (is => 'ro', isa => 'Str', default => DEFAULT_SYMBOL_LIST);

sub BUILD
{
    my ($self) = @_;

    $self->{quote} = Finance::Quote->new();
    $self->{quote}->timeout(6);
}

sub cache_ttl { 60 };

sub get_vars
{
    my ($self) = @_;

    say 'Getting data for headline symbol: ', $self->headline_symbol();
    my $headline_data_ref = $self->_strip_symbol(
                            $self->_fetch(
                            $self->headline_symbol() ));

    say 'Getting data for symbols: ', $self->symbol_list();
    my $data_ref = $self->_reformat(
                   $self->_fetch(
                   split(q{,}, $self->symbol_list()) ));

    my %page_data =
    (
        headline_data => $headline_data_ref,
        data => $data_ref,
    );
    return \%page_data;
}

sub _fetch
{
    my ($self, @symbols) = @_;

    return $self->{quote}->fetch(QUOTE_LOCATION, @symbols);
}

sub _reformat
{
    my ($self, %data_ref) = @_;

    my %new_data;

    while (my ($key, $value) = each %data_ref)
    {
        my ($symbol, $subkey) = split(chr(0x1c), $key);
        $new_data{$symbol}->{$subkey} = $value;
    }

    return \%new_data;
}

sub _strip_symbol
{
    my ($self, %data_ref) = @_;

    my %new_data;

    while (my ($key, $value) = each %data_ref)
    {
        my (undef, $subkey) = split(chr(0x1c), $key);
        $new_data{$subkey} = $value;
    }

    return \%new_data;
}

1;


Flowcase/lib/Flowcase/Pages/Weather.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase::Pages::Weather;

use common::sense;

use JSON::PP;
use LWP::UserAgent;
use Moose;

has 'apikey' => (is => 'ro', isa => 'Str');
has 'place' => (is => 'ro', isa => 'Str');

sub BUILD
{
    my ($self) = @_;

    $self->{ua} = LWP::UserAgent->new();
    $self->{json} = JSON::PP->new();
}

sub cache_ttl { 240 };

sub get_vars
{
    my ($self) = @_;

    return { %{$self->_get_forecast_vars()},
             %{$self->_get_conditions_vars()} };
}

sub get_url
{
    my ($self, $url) = @_;
    say "Getting URL [$url]";
    my $response = $self->{ua}->get($url);
    return $response->is_success() ?
           $response->decoded_content() :
           undef;
}

sub _get_forecast_vars
{
    my ($self) = @_;

    my $url = 'http://api.wunderground.com/' .
              'api/' . $self->{apikey} . '/' .
              "forecast/q/$self->{place}.json";
    my $json = $self->get_url($url);

    my $forecast = $self->{json}->decode($json);
    my $days = $forecast->{forecast}->{simpleforecast}->{forecastday};

    my %vars = (days => []);
    my $count = 0;
    for my $day (@$days)
    {
        push(@{$vars{days}},
        {
            weekday => $day->{date}->{weekday},
            weather => $day->{conditions},
            high    => $day->{high}->{fahrenheit},
            low     => $day->{low}->{fahrenheit},
        });

        last if ++$count >= 4;
    }

    return \%vars;
}

sub _get_conditions_vars
{
    my ($self) = @_;

    my $url = 'http://api.wunderground.com/' .
              'api/' . $self->{apikey} . '/' .
              "conditions/q/$self->{place}.json";
    my $json = $self->get_url($url);

    my $weather = $self->{json}->decode($json);
    my $cur = $weather->{current_observation};

    my %vars =
    (
        feels_like    => $cur->{feelslike_f},
        humidity      => $cur->{relative_humidity},
        location      => $cur->{display_location}->{full},
        temperature   => $cur->{temp_f},
        weather       => $cur->{weather},
        wind_dir      => $cur->{wind_dir},
        wind_gust_mph => sprintf('%.1f', $cur->{wind_gust_mph}),
        wind_mph      => sprintf('%.1f', $cur->{wind_mph}),
    );

    return \%vars;
}

1;


Flowcase/lib/Flowcase.pm
# Copyright 2012, 2013 Colin Wetherbee
#
# This file is part of Flowcase.
#
# Flowcase is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Flowcase is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Flowcase.  If not, see <http://www.gnu.org/licenses/>.

package Flowcase;

use common::sense;

use Getopt::Long;
use Moose;

use Flowcase::WebHandler;

# Default listen IP address.
use constant DEFAULT_LISTEN_ADDRESS => '0.0.0.0';

# Default delay between page transitions, in seconds.
use constant DEFAULT_PAGE_DELAY => 60;

# Default port on which to listen for web requests.
use constant DEFAULT_LISTEN_PORT => 8958;

# Default path for page model modules.
use constant DEFAULT_MODULE_PATH => 'Flowcase::Pages';

# Command-line options definition.
use constant OPTS =>
[qw(
    help
    address|a=s config|c=s% delay|d=i module_path|m=s
    port|p=i tmpl_dir|t=s exclude|x=s
)];

=head1 NAME

Flowcase - A web-based slideshow of any content you want!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Run the Flowcase application.

    use Flowcase;

    my $flowcase = Flowcase->new();
    my $exit_code = $flowcase->start();
    exit $exit_code;

=head1 SUBROUTINES/METHODS

=cut

our $_singleton;
sub get_instance
{
    return $_singleton;
}

has 'listen_address' =>
    ( isa => 'Str', is => 'rw', default => DEFAULT_LISTEN_ADDRESS );
has 'page_delay' =>
    ( isa => 'Num', is => 'rw', default => DEFAULT_PAGE_DELAY );
has 'listen_port' =>
    ( isa => 'Num', is => 'rw', default => DEFAULT_LISTEN_PORT );
has 'module_path' =>
    ( isa => 'Str', is => 'rw', default => DEFAULT_MODULE_PATH );
has 'tmpl_dir' =>
    ( isa => 'Str', is => 'rw' );
has 'exclude' =>
    ( isa => 'Str', is => 'rw' );
has 'config' =>
    ( isa => 'HashRef', is => 'rw');

=head2 start()

Start Flowcase.  Parse command-line options and begin the main loop.

=cut
sub start
{
    die('Too many instances of ' . __PACKAGE__) if $_singleton;
    my ($self) = @_;

    $_singleton = $self;

    $self->parse_opts();

    say 'Options to be used:';
    say ">> Listen address: $self->{listen_address}";
    say ">> Listen port: $self->{listen_port}";
    say ">> Module path: $self->{module_path}";
    say ">> Page delay: $self->{page_delay}";
    say ">> Template directory: $self->{tmpl_dir}";
    say '>> Pages to exclude:';
    say '    (none)' if scalar(@{$self->{exclude}}) == 0;
    foreach (sort { $a cmp $b } @{$self->{exclude}})
    {
        say "    $_";
    }
    say '>> Config options:';
    say '    (none)' if scalar(keys %{$self->{config}}) == 0;
    foreach (sort { $a cmp $b } keys %{$self->{config}})
    {
        say "    '$_' => '$self->{config}->{$_}'";
    }

    Flowcase::WebHandler->new()->start();

    # Return a UNIX-style exit code.
    return 0;
}

sub _usage
{
    my ($self, $errors_ref) = @_;

    if ($errors_ref && scalar(@$errors_ref) > 0)
    {
        say STDERR "ERROR: Missing or invalid value for option(s):";
        say "    $_" for (@$errors_ref);
        say q{};
    }

    say 'Usage: flowcase -t directory [-d number]';
    say 'Required arguments:';
    say '    -t directory         Directory that contains HTML templates';
    say 'Optional arguments:';
    say '    -a address           IP address on which to listen for web';
    say '                         requests [', DEFAULT_LISTEN_ADDRESS, ']';
    say '    -d number            Delay, in seconds, between page';
    say '                         transitions [', DEFAULT_PAGE_DELAY, ']';
    say '    -m path              Perl module path for page models';
    say '                         [', DEFAULT_MODULE_PATH, ']';
    say '    -p port              Port on which to listen for web requests';
    say '                         [', DEFAULT_LISTEN_PORT, ']';
    say '    -x pages             Comma-separated list of page files to ';
    say '                         exclude';

    exit($errors_ref ? -1 : 0);
}

=head2 parse_opts()

Parse command-line options and store them (or defaults) in $self.

=cut
sub parse_opts
{
    my ($self) = @_;

    my %opts;
    GetOptions(\%opts, @{+OPTS});

    $self->_usage() if $opts{help};

    my @errors;

    if (defined($opts{address}))
    {
        $self->listen_address($opts{address});
    }

    if (defined($opts{delay}))
    {
        if ($opts{delay} !~ /^\d+$/ || $opts{delay} < 1)
        {
            push(@errors, 'Delay must be 1 or greater');
        }
        else
        {
            $self->page_delay($opts{delay});
        }
    }

    if (defined($opts{module_path}))
    {
        $self->module_path($opts{module_path});
    }

    if (defined($opts{port}))
    {
        if ($opts{port} =~ /^\d+$/)
        {
            if ($opts{port} <= 0 || $opts{port} >= 65536)
            {
                push(@errors, 'Port must be between 0 and 65536, exclusive');
            }
            elsif ($> != 0 && $opts{port} <= 1024)
            {
                push(@errors, 'Non-root users are restricted to high ports');
            }
            else
            {
                $self->listen_port($opts{port});
            }
        }
        else
        {
            push(@errors, 'Port must be numeric');
        }
    }

    if (defined($opts{tmpl_dir}))
    {
        if (-d $opts{tmpl_dir})
        {
            $self->tmpl_dir($opts{tmpl_dir});
        }
        else
        {
            push(@errors, 'Template directory must exist');
        }
    }
    else
    {
        push(@errors, 'Template directory must be provided');
    }

    $self->{exclude} = $opts{exclude} ? [ split(q{,}, $opts{exclude}) ] : [];

    if (defined($opts{config}))
    {
        $self->{config} = $opts{config};
    }

    $self->_usage(\@errors) if (scalar @errors > 0)
}

=head1 AUTHOR

Colin Wetherbee, C<< <cww at cpan.org> >>

=head1 BUGS

Please report any bugs using the GitHub issue tracker for this project.

L<https://github.com/cww/flowcase/issues>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Flowcase

=head1 LICENSE AND COPYRIGHT

Please see the LICENSE file included in this distribution for the terms under
which this software is licensed.

Additionally, each source file includes a copyright notice and brief licensing
terms.

=cut

1;
